source/extensions/transport_sockets/tls/cert_validator/default_validator.cc

```diff
@@ -276,6 +304,78 @@ Envoy::Ssl::ClientValidationStatus DefaultCertValidator::verifyCertificate(
   return validated;
 }
 
+ValidationResults DefaultCertValidator::doVerifyCertChain(
+    STACK_OF(X509) & cert_chain, Ssl::ValidateResultCallbackPtr /*callback*/,
+    Ssl::SslExtendedSocketInfo* ssl_extended_info,
+    const Network::TransportSocketOptions* transport_socket_options, SSL_CTX& ssl_ctx,
+    absl::string_view ech_name_override, bool is_server, uint8_t current_tls_alert) {
+  if (sk_X509_num(&cert_chain) == 0) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(
+          Envoy::Ssl::ClientValidationStatus::NotValidated);
+    }
+    stats_.fail_verify_error_.inc();
+    std::string error = "verify cert failed: empty cert chain";
+    ENVOY_LOG(debug, error);
+    return {ValidationResults::ValidationStatus::Failed, absl::nullopt, error};
+  }
+  X509* leaf_cert = sk_X509_value(&cert_chain, 0);
+  ASSERT(leaf_cert);
+  if (verify_trusted_ca_) {
+    X509_STORE* verify_store = SSL_CTX_get_cert_store(&ssl_ctx);
+    ASSERT(verify_store);
+    bssl::UniquePtr<X509_STORE_CTX> ctx(X509_STORE_CTX_new());
+    if (!ctx || !X509_STORE_CTX_init(ctx.get(), verify_store, leaf_cert, &cert_chain) ||
+        // We need to inherit the verify parameters. These can be determined by
+        // the context: if it's a server it will verify SSL client certificates or
+        // vice versa.
+        !X509_STORE_CTX_set_default(ctx.get(), is_server ? "ssl_client" : "ssl_server") ||
+        // Anything non-default in "param" should overwrite anything in the ctx.
+        !X509_VERIFY_PARAM_set1(X509_STORE_CTX_get0_param(ctx.get()),
+                                SSL_CTX_get0_param(&ssl_ctx)) ||
+        // ClientHelloOuter connections use a different name.
+        (!ech_name_override.empty() &&
+         !X509_VERIFY_PARAM_set1_host(X509_STORE_CTX_get0_param(ctx.get()),
+                                      ech_name_override.data(), ech_name_override.size()))) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_X509_LIB);
+      if (ssl_extended_info) {
+        ssl_extended_info->setCertificateValidationStatus(
+            Envoy::Ssl::ClientValidationStatus::Failed);
+      }
+      stats_.fail_verify_error_.inc();
+      std::string error = "verify cert failed: init and setup X509_STORE_CTX";
+      ENVOY_LOG(debug, error);
+      return {ValidationResults::ValidationStatus::Failed, absl::nullopt, error};
+    }
+    int ret = X509_verify_cert(ctx.get());
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(
+          ret == 1 ? Envoy::Ssl::ClientValidationStatus::Validated
+                   : Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+
+    if (ret <= 0) {
+      stats_.fail_verify_error_.inc();
+      const std::string error =
+          absl::StrCat("verify cert failed: ", Utility::getX509VerificationErrorInfo(ctx.get()));
+      ENVOY_LOG(debug, error);
+      if (allow_untrusted_certificate_) {
```

Comment:
If you're going to be allowing all certificates anyway, it's kind of odd to still run the verification. Or do you want the logging and `ssl_extended_info` business?

Reply:
This is following what doSynchronousVerifyCertChain() is already doing. I believe it is for retaining verify result into ssl_extended_info.